examples/nlp/gpt2_text_generation_with_kerasnlp.py

```diff
@@ -0,0 +1,363 @@
+    output of `GPT2Backbone` by embedding matrix to generate logits over
+    vocab tokens.
+"""
+
+"""
+## Finetune on Reddit dataset.
+
+Now you have the knowledge of the GPT-2 model from KerasNLP, you can take one
+step further to finetune the model so that it generates text in a specific
+style, short or long, strict or casual. In this tutorial, we will use reddit
+dataset for example.
+"""
+
+import tensorflow_datasets as tfds
+
+reddit_ds = tfds.load("reddit_tifu", split="train", as_supervised=True)
+
+"""
+Let's take a look inside sample data from the reddit TensorFlow Dataset. There
+are two features:
+
+- **__document__**: text of the post.
+- **__title__**: the title.
+
+"""
+
+for document, title in reddit_ds:
+    print(document.numpy())
+    print(title.numpy())
+    break
+
+"""
+In our case, we are performing next word prediction in a language model, so we
+only need the 'document' feature.
+"""
+
+train_ds = (
+    reddit_ds.map(lambda document, _: document)
+    .batch(32)
+    .cache()
+    .prefetch(tf.data.AUTOTUNE)
+)
+
+"""
+Now you can finetune the model using the familiar *fit()* function. Note that
+`preprocessor` will be automatically called inside `fit` method since
+`GPT2CausalLM` is a `keras_nlp.models.Task` instance.
+
+This step takes quite a bit of GPU memory and a long time if we were to train
+it all the way to a fully trained state. Here we just use part of the dataset
+for demo purposes.
+"""
+
+train_ds = train_ds.take(500)
+num_epochs = 1
+
+learning_rate = keras.optimizers.schedules.PolynomialDecay(
+    5e-5,
+    decay_steps=train_ds.cardinality() * num_epochs,
+    end_learning_rate=0.0,
+)
+loss = keras.losses.SparseCategoricalCrossentropy(from_logits=True)
+gpt2_lm.compile(
+    optimizer=keras.optimizers.Adam(learning_rate),
+    loss=loss,
+    weighted_metrics=["accuracy"],
+)
+
+gpt2_lm.fit(train_ds, epochs=num_epochs)
+
+"""
+After finetuning is finished, you can again generate text using the same
+*generate()* function. This time, the text will be closer to Reddit writing
+style, and the generated length will be close to our preset length in the
+training set.
+"""
+
+start = time.time()
+
+output = gpt2_lm.generate("I like basketball", max_length=200)
+print("\nGPT-2 output:")
+print(output)
+
+end = time.time()
+print(f"TOTAL TIME ELAPSED: {end - start:.2f}s")
+
+"""
+## Into the Sampling Method
+
+In KerasNLP, we offer a few sampling methods, e.g., contrastive search,
+Top-K and beam sampling. By default our `GPT2CausalLM` uses contrastive
+search, but you can choose your own sampling method.
+
+Much like optimizer and activations, there are two ways to specify your custom
+sampler:
+
+- Use a string identifier, such as "top_k", you are using the default
+configuration via this way.
+- Pass a `keras_nlp.samplers.Sampler` instance, you can use custom configuration
+via this way.
+"""
+
+# Use a string identifier.
+gpt2_lm.compile(sampler="top_k")
```

Comment:
maybe use some samplers that will actually show noticeable difference here, might be cool for a reader

`sampler="greedy"`
`sampler="random"`
`sampler=keras_nlp.samplers.TopK(k=10)`

And make some comments about how greedy will tend to repeat itself, random sampling can tend to choose weird outputs, and top-k will be somewhere in the middle?